presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -93,101 +103,200 @@ public void stop()
+                        })
+                        .forEach(stageDomains -> newDynamicFiltersBuilder.put(stageDomains.getKey(), union(stageDomains.getValue())));
             }
 
-            Map<SourceDescriptor, Domain> newDynamicFilters = newDynamicFiltersBuilder.build();
+            Map<DynamicFilterId, Domain> newDynamicFilters = newDynamicFiltersBuilder.build();
             if (!newDynamicFilters.isEmpty()) {
                 addDynamicFilters(queryId, newDynamicFilters);
             }
         }
     }
 
-    public Supplier<TupleDomain<ColumnHandle>> createDynamicFilterSupplier(QueryId queryId, List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles)
+    @VisibleForTesting
+    Optional<Domain> getSummary(QueryId queryId, DynamicFilterId filterId)
     {
-        Map<DynamicFilterId, ColumnHandle> sourceColumnHandles = extractSourceColumnHandles(dynamicFilters, columnHandles);
-
-        return () -> dynamicFilters.stream()
-                .map(filter -> getSummary(queryId, filter.getId())
-                        .map(summary -> translateSummaryToTupleDomain(filter.getId(), summary, sourceColumnHandles)))
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .reduce(TupleDomain.all(), TupleDomain::intersect);
+        return getSummary(dynamicFilterSummaries.get(queryId).get(filterId));
     }
 
-    @VisibleForTesting
-    Optional<Domain> getSummary(QueryId queryId, DynamicFilterId filterId)
+    private Optional<Domain> getSummary(ListenableFuture<Domain> future)
     {
-        return Optional.ofNullable(dynamicFilterSummaries.get(SourceDescriptor.of(queryId, filterId)));
+        if (!future.isDone()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(getDone(future));
     }
 
-    private synchronized Map<QueryId, Supplier<List<StageDynamicFilters>>> getQueries()
+    private Map<QueryId, Supplier<List<StageDynamicFilters>>> getDynamicFilterSuppliers()
     {
-        return ImmutableMap.copyOf(queries);
+        return ImmutableMap.copyOf(dynamicFilterSuppliers);
     }
 
-    private synchronized void addDynamicFilters(QueryId queryId, Map<SourceDescriptor, Domain> dynamicFilters)
+    private synchronized void addDynamicFilters(QueryId queryId, Map<DynamicFilterId, Domain> dynamicFilters)
     {
-        // query might have been removed while we collected dynamic filters asynchronously
-        if (!queries.containsKey(queryId)) {
+        Map<DynamicFilterId, SettableFuture<Domain>> dynamicFilterFutures = dynamicFilterSummaries.get(queryId);
+        if (dynamicFilterFutures == null) {
+            // query might have been removed while filters were collected asynchronously
             return;
         }
 
-        dynamicFilterSummaries.putAll(dynamicFilters);
+        dynamicFilters.forEach((filter, domain) -> dynamicFilterFutures.get(filter).set(domain));
```

Comment:
Consider verifying that the same future is not set twice:
```suggestion
        dynamicFilters.forEach((filter, domain) -> {
            SettableFuture<Domain> future = requireNonNull(dynamicFilterFutures.get(filter), "Future not found");
            verify(future.set(domain), "Same future set twice");
        });
```